Remove references of deprecated const IL Opcode

wbh123456 opened the pull request:
Remove all references to the IL Opcodes in the `const` category listed in #2657, namely,  `buconst`, `iuconst`, `luconst`, `cconst`. 
Also removed `TR::Node::cconst` 

*Anything in `SwitchAnalyzer.cpp` as they will be removed explicitly in #3983*

Issue: #2657 
Signed-off-by: Bohao(Aaron) Wang <aaronwang0407@gmail.com>

andrewcraik commented:
Before removal of opcodes from optimizations happens we need to ensure no downstream projects are generating any instances of the opcode. I would go so far as to suggest that the tree verifier that runs during compilation (the old one as well as the new non-default one) are both modified to fatal assert if the opcode is seen once we believe all sources of the opcode have been eliminated. That way no one will add any code to generate these deprecated opcodes and removal can progress. Without some kind of always on assert and verification of downstream consumers I worry we may be adding bugs (either functional or performance) by removing handling...

0xdaryl commented:
Please check the linter failures.

wbh123456 commented on compiler/x/codegen/BinaryEvaluator.cpp:
```diff
@@ -1784,8 +1783,7 @@ TR::Register *OMR::X86::TreeEvaluator::csubEvaluator(TR::Node *node, TR::CodeGen
       }
 
    TR::Register * testRegister = secondChild->getRegister();
-   if (secondChild->getOpCodeValue() == TR::cconst &&
```
This whole if statement can be removed.  The condition is always `false` as the `TR::cconst` no longer exists

andrewcraik replied:
I think this may need to be changed to sconst - I don't think it can be completely removed.

wbh123456 replied:
Fixed in c215804

wbh123456 commented on compiler/x/codegen/BinaryEvaluator.cpp:
```diff
@@ -1252,8 +1252,7 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
       }
 
    TR::Register * testRegister = secondChild->getRegister();
-   if (secondChild->getOpCodeValue() == TR::cconst &&
```
This whole if statement can be removed. The condition is always false as the TR::cconst no longer exists

andrewcraik replied:
Yeah I think this should be sconst with appropriate updates - I don't think the whole case can be removed

wbh123456 replied:
Fixed in c215804.

wbh123456 commented:
@vijaysun-omr This PR is good for a review.

vijaysun-omr commented:
@genie-omr build all

wbh123456 commented:
> @genie-omr build all

I suspect the build failure was caused by the linking error of `cconstEvaluator` because I removed the definition of it but did not remove its declaration. In the new fore-push, I recovered the definition of `cconstEvaluator`. Could you build this PR again?

vijaysun-omr commented:
@genie-omr build all

vijaysun-omr commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -4326,9 +4326,7 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
-         if (secondOp == TR::su2i ||
-             (secondOp == TR::iuconst               &&
-              secondChild->getUnsignedInt() <= USHRT_MAX))
+         if (secondOp == TR::su2i || secondChild->getUnsignedInt() <= USHRT_MAX)
```
Is this logically equivalent to what was there before the change ? It seems to me that you should have omitted the ```secondChild->getUnsignedInt() <= USHRT_MAX``` test on the basis that ```secondOp == TR::iuconst``` should be false.

wbh123456 replied:
Yes, this was a mistake. `(secondOp == TR::iuconst&&secondChild->getUnsignedInt() <= USHRT_MAX)` should all be omitted as it will always be `false`. Fix in the new force-push.

vijaysun-omr commented on compiler/x/codegen/BinaryEvaluator.cpp:
```diff
@@ -1252,8 +1252,7 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
    TR::Register * testRegister = secondChild->getRegister();
-   if (secondChild->getOpCodeValue() == TR::cconst &&
-       testRegister == NULL         &&
+   if (testRegister == NULL         &&
```
Is this logic what you intended here ? I ask because the opcode value not being ```cconst``` suggests that this if condition ought not to be entered.

wbh123456 replied:
I believe you are reviewing the outdated version of this. I have deleted this whole if statement.

wbh123456 commented:
Did some code simplifying in `x/codegen/BinaryEvaluator.cpp` in the latest [commit].(https://github.com/eclipse/omr/pull/4033/commits/0932037936ea4310a07be2740c897d1435ae1fbf)

vijaysun-omr commented:
@genie-omr build all

vijaysun-omr commented:
@andrewcraik are you okay with this PR ?

andrewcraik commented:
I don't think we should merge this yet. OpenJ9 has at least one place that can still generate `cconst` nodes (https://github.com/eclipse/openj9/blob/3306c99a7dfcc5b033e98dad7ff810a3e546b3c3/runtime/compiler/optimizer/SequentialStoreSimplifier.cpp#L1777) and other downstream projects may have some as well. Until downstream projects have cleaned up their use this change would break them by removing the evaluators they are likely using and optimizations they may be relying upon. We should help those projects clean-up their usage before we remove the opcode from OMR.

vijaysun-omr commented:
@wbh123456 ping ?

wbh123456 commented:
> @wbh123456 ping ?

Still working on removing unsigned opcodes from OpenJ9. Once that's done, we can merge this.

vijaysun-omr commented:
@wbh123456 any update on this ?

wbh123456 commented:
> @wbh123456 any update on this ?

This PR is still work in progress as it we need to stop generating all `const` IL Opcode before we can remove the references to them.  https://github.com/eclipse/omr/pull/4347.

wbh123456 commented:
There is one more reference to `Node::cconst()` in OpenJ9. It is going to be removed in https://github.com/eclipse/openj9/pull/7909. This PR will be ready once that PR is merged.

wbh123456 commented:
This PR is ready for a review. @andrewcraik @vijaysun-omr

Leonardo2718 commented:
@genie-omr build all

vijaysun-omr commented:
The logic after the cleanup looks right to me. Given that the downstream project dependencies have now been eliminated, I think this can go in if @andrewcraik gives his approval.

andrewcraik commented on compiler/optimizer/OMRSimplifierHelpers.cpp:
```diff
@@ -346,8 +346,6 @@ TR::Node *foldRedundantAND(TR::Node * node, TR::ILOpCodes andOpCode, TR::ILOpCod
       {
       switch(constOpCode)
          {
-         case TR::sconst: case TR::cconst:
```
I don't think this is right - didn't you mean to just delete the `cconst` case - the sconst is not being removed right?

wbh123456 replied:
Sorry, that was a mistake, only `cconst` should be deleted. Fixed in c215804

andrewcraik commented:
There are two removals in the codegen that don't seem right - ccsont should be mapped to sconst for these I believe @wbh123456 please review and comment. Once we conclude this discussion I think this will be ready.

wbh123456 commented:
> There are two removals in the codegen that don't seem right - ccsont should be mapped to sconst for these I believe @wbh123456 please review and comment. Once we conclude this discussion I think this will be ready.

I agree that those two removals are not correct. I have reverted back the if statement and mapped `cconst` to `sconst` in c215804. I don't think there are anything else needed to be updated, do you agree on that?

vijaysun-omr commented:
@genie-omr build all

andrewcraik commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -4340,24 +4338,6 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
                   dumpOptDetails(s->comp(), "Integer Compare Narrower: found both children c2i in method %s\n", s->comp()->signature());
                   }
                }
-            else if (secondOp == TR::iuconst)
```
Why is this removed? shouldn't it be an iconst?

wbh123456 replied:
This is a mistake. The `TR::iuconst` in this `IntComp` pattern matching should be mapped to `TR::iconst` in this case. I have fixed this problem in cdd352b

andrewcraik commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -4408,9 +4388,7 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
       else if (firstOp == TR::b2i && firstChild->getReferenceCount() == 1)
          {
-         if (secondOp == TR::b2i ||
-             (secondOp == TR::iuconst               &&
```
shouldn't this be an iconst with a getUnsignedInt to test against SCHAR_MAX?

wbh123456 replied:
Fixed.

andrewcraik commented on compiler/optimizer/OMRSimplifierHandlers.cpp:
```diff
@@ -4424,25 +4402,6 @@ static void unsignedIntCompareNarrower(TR::Node * node, TR::Simplifier * s, TR::
                   dumpOptDetails(s->comp(), "Integer Compare Narrower: found both children b2i in method %s\n", s->comp()->signature());
                   }
                }
-            else if (secondOp == TR::iuconst)
```
again iconst with appropriate adaptations rather than removal...

wbh123456 replied:
Fixed

wbh123456 commented:
@andrewcraik I have fixed the false removal of `TR::iuconst` in cdd352b. I have mapped the `TR::iuconst` to `TR::iconst`. In addition, I have also reverted back the test against `SCHAR_MAX` as it is still valid in with `TR:iconst`

wbh123456 commented:
@andrewcraik I have fixed `iuconst` mapping problems in `OMRSimplifierHandlers.cpp`.  I have also scanned all the mappings of deprecated unsigned const opcodes, which all look appropriate to me.  This PR is ready for another review.

andrewcraik commented on compiler/optimizer/VPHandlers.cpp:
```diff
@@ -10748,7 +10748,7 @@ static TR::Node *constrainCmplessthan(OMR::ValuePropagation *vp, TR::Node *node,
          {
          vp->removeChildren(node);
-         TR::ILOpCodes op = /*isUnsigned ? TR::iuconst : */TR::iconst;
+         TR::ILOpCodes op = TR::iconst;
```
can you just put iconst directly into the recreate to save a needless local?

wbh123456 replied:
Fixed in e4cb262

andrewcraik commented on compiler/x/codegen/BinaryEvaluator.cpp:
```diff
@@ -1252,49 +1252,49 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
+      if (value >= 0 && value <= 127)
          {
-         if (value >= 0 && value <= 127)
+         if (value == 1)
```
I'm not sure this is right - the sconst is read as unsigned - isn't it possible for the value to be between 128 and 255 inclusive. I know the node is now sconst but that doesn't change the range when the const is still read as a uint16_t...

andrewcraik commented on compiler/x/codegen/BinaryEvaluator.cpp:
```diff
@@ -1252,49 +1252,49 @@ TR::Register *OMR::X86::TreeEvaluator::caddEvaluator(TR::Node *node, TR::CodeGen
+   if (secondChild->getOpCodeValue() == TR::sconst &&
+      testRegister == NULL         &&
+      performTransformation(comp, "O^O CaddEvaluator: checking that the store has not happened yet. Target register:  %x\n", testRegister))
+   {
```
Something weird happened here in the formatting - can you sort the diff to make more sense - looks like the if body got outdented?

wbh123456 replied:
I am not sure what happens here. The formatting looks exactly the same on both sides. Git diff tool on my local repo also can't detect a difference except from line `1255`.  Lines from `1256` to `1297` should not have changed at all. I have tried reverted the change in line `1255`, it did not work.

andrewcraik commented on compiler/z/codegen/UnaryEvaluator.cpp:
```diff
@@ -117,13 +117,8 @@ TR::Register *
-      TR_ASSERT(1, "Should not have bconst node in the final il-tree!\n");
-      generateLoad32BitConstant(cg, node, node->getByte(), tempReg, true);
-      }
+   TR_ASSERT(1, "Should not have bconst node in the final il-tree!\n");
```
@fjeremic is this assert even valid? this should be removed shouldn't it (in the context of OMR)

wbh123456 replied:
@fjeremic Ping

fjeremic replied:
This assert seems bogus. `bconst` trees should be valid in the OMR context. I suspect this may be some Java-ism leaking over. This evaluator looks correct. We should remove the assert.

wbh123456 replied:
I have deleted the assert in c56d3c5.

andrewcraik commented:
Opcode transform and handling now looks right - just a few smaller things to look at and we can look to merge this.

andrewcraik commented on compiler/optimizer/LoopReducer.cpp:
```diff
@@ -1811,7 +1811,7 @@ TR_Arraytranslate::checkBreak(TR::Block * breakBlock, TR::Node * breakNode, TR::
       //      add or subtract 1 from the value and still represent it in 2**16 bits for a default value
       //      if we need to generate a table.
-      dumpOptDetails(comp(), "...break tree does not have bconst/cconst/iconst, or not in range - no arraytranslate reduction\n");
+      dumpOptDetails(comp(), "...break tree does not have bconst/iconst, or not in range - no arraytranslate reduction\n");
```
shouldn't sconst be in here rather than deleting cconst?

wbh123456 replied:
The if statement is not checking `cconst` nor `bconst`. Maybe this `dumpOptDetails()` is not up to date. Should I just delete both `cconst` and `bconst`?

andrewcraik replied:
LoopReducer is not heavily used anymore so I would say if the condition is only checking for iconst just update the details to say as much

wbh123456 replied:
Fixed.

0xdaryl commented:
@genie-omr build all
